integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -38,81 +58,63 @@ public fun <T> CoroutineScope.future(
-            } catch (t: Throwable) {
-                setException(t)
-            }
-        }
-    }
-    override fun interruptTask() { deferred.cancel() }
-}
-
 /**
- * Converts this listenable future to an instance of [Deferred].
- * It is cancelled when the resulting deferred is cancelled.
+ * Returns a [Deferred] that is completed or failed by `this` [ListenableFuture].
+ *
+ * Completion is non-atomic between the two promises.
+ *
+ * When either promise successfully completes, it will attempt to synchronously complete its
+ * counterpart with the same value. This will succeed barring a race with cancellation.
+ *
+ * When either promise completes with an Exception, it will attempt to synchronously complete its
+ * counterpart with the same Exception. This will succeed barring a race with cancellation.
+ *
+ * Cancellation is propagated bidirectionally.
+ *
+ * When `this` [ListenableFuture is successfully cancelled - meaning [ListenableFuture.cancel]
+ * returned `true` - it will synchronously cancel the returned [Deferred]. This can only race with
+ * cancellation of the returned `Deferred`, so the `Deferred` will always be put into its
+ * "cancelling" state and (barring uncooperative cancellation) _eventually_ reach its "cancelled"
+ * state when either promise is successfully cancelled.
+ *
+ * When the returned [Deferred] is cancelled [ListenableFuture.cancel] will be synchronously called
+ * on `this` [ListenableFuture]. This will attempt to cancel
+ * the `Future`, though cancellation may not succeed and the [ListenableFuture] may complete in a
+ * non-cancelled terminal state.
+ *
+ * This method may receive and suppress the `true` return value from [ListenableFuture.cancel], if
+ * the task is cancelled via the [Deferred] reference to it. This is unavoidable, so make sure
+ * no idempotent work is performed by any users of the [ListenableFuture] reference to the task
+ * if they're the thread that succeeded in cancelling the task - it may not get done.
+ *
+ * This is inherently a race. See [Future.cancel] for a description of `Future` cancellation
+ * semantics. See [Job] for a description of coroutine cancellation semantics.
  */
 public fun <T> ListenableFuture<T>.asDeferred(): Deferred<T> {
-    // Fast path if already completed
-    if (isDone) {
+    // Fast path if already completed.
+    // TODO: There's a faster fast path - InternalFutureFailureAccess - in Guava >= 27.0
+    if (isDone && !isCancelled) {
         return try {
-            @Suppress("UNCHECKED_CAST")
-            CompletableDeferred(get() as T)
-        } catch (e: Throwable) {
-            // unwrap original cause from ExecutionException
-            val original = (e as? ExecutionException)?.cause ?: e
-            CompletableDeferred<T>().also { it.completeExceptionally(original) }
+            // getUninterruptibly() avoids one volatile read over Futures.getDone().
+            CompletableDeferred(Uninterruptibles.getUninterruptibly(this))
```

Comment:
Guava has a bad reputation with `@Beta` API compatibility. It's better to use stable API, especially if it only costs one volatile read

Reply:
This method is not `@Beta`:

https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/Uninterruptibles.java#L232

Reply:
Good to know, apparently we use a slightly outdated version of Guava